packages/analytics-js-plugins/src/deviceModeTransformation/index.ts

```diff
@@ -17,25 +39,159 @@ const DeviceModeTransformation = (): ExtensionPlugin => ({
   },
   transformEvent: {
+    init(
+      state: ApplicationState,
+      pluginsManager: IPluginsManager,
+      httpClient: IHttpClient,
+      storeManager: IStoreManager,
+      errorHandler?: IErrorHandler,
+      logger?: ILogger,
+    ) {
+      const writeKey = state.lifecycle.writeKey.value as string;
+      httpClient.setAuthHeader(writeKey);
+
+      const eventsQueue = new RetryQueue(
+        // adding write key to the queue name to avoid conflicts
+        `${QUEUE_NAME}_${writeKey}`,
+        DEFAULT_TRANSFORMATION_QUEUE_OPTIONS,
+        (item: TransformationQueueItemData, done: DoneCallback) => {
+          httpClient.getAsyncData({
+            url: `${state.lifecycle.dataPlaneUrl.value}/transform`,
+            options: {
+              method: 'POST',
+              data: getDMTDeliveryPayload(item.payload) as string,
+              sendRawData: true,
+            },
+            isRawResponse: true,
+            timeout: REQUEST_TIMEOUT_MS,
+            callback: (result, details) => {
+              // null means item will not be requeued
+              const queueErrResp = isErrRetryable(details) ? details : null;
+
+              pluginsManager.invokeSingle(
+                'transformEvent.sendDataToDestination',
+                state,
+                pluginsManager,
+                result,
+                details?.xhr?.status,
+                item.event,
+                errorHandler,
+                logger,
+              );
+              done(queueErrResp, result);
+            },
+          });
+        },
+        storeManager,
+        MEMORY_STORAGE,
+      );
+
+      return eventsQueue;
+    },
+
     enqueue(
+      state: ApplicationState,
+      eventsQueue: IQueue,
+      event: RudderEvent,
+      destinations: Destination[],
+    ) {
+      const destinationIds = destinations.map(d => d.id);
+      const payload = createPayload(event, destinationIds, state.session.authToken.value);
+      eventsQueue.addItem({ event, payload } as TransformationQueueItemData);
+    },
+
+    sendDataToDestination(
       state: ApplicationState,
       pluginsManager: IPluginsManager,
+      result: any,
+      status: number,
       event: RudderEvent,
-      destination: Destination,
       errorHandler?: IErrorHandler,
       logger?: ILogger,
     ) {
-      // TODO: Implement DMT logic here
-
-      // TODO: for now this is a pass through
-      pluginsManager.invokeSingle(
-        'destinationsEventsQueue.enqueueEventToDestination',
-        state,
-        event,
-        destination,
-        errorHandler,
-        logger,
-      );
+      const NATIVE_DEST_EXT_POINT = 'destinationsEventsQueue.enqueueEventToDestination';
+      const ACTION_TO_SEND_UNTRANSFORMED_EVENT = 'Sending untransformed event';
+      const ACTION_TO_DROP_EVENT = 'Dropping the event';
+      const destinations = state.nativeDestinations.dmtEnabledDestinations.value;
```

Comment:
This state value is updated when a new event is processed in the destination events queue while a DMT request is in progress. So, it'll lead to race condition behavior. Let's change the logic to pass the destinations across the control flow instead of the state.

Reply:
I have tried that, once you push the destinations in the queue, the initialized instance somehow loses the functions attached to it. instance.page() becomes undefined

Reply:
We can see why it's not working, but this approach is incorrect. 
Instead of destination instance objects, you can pass around their IDs and resolve the instance from the state.

CC @bardisg

Reply:
Done